collect_app/src/test/java/org/odk/collect/android/widgets/AudioWidgetTest.java

```diff
@@ -1,117 +1,351 @@
+        assertThat(audioHelper.isMediaStopped, equalTo(true));
     }
 
-    @Override
-    public Object createBinaryData(StringData answerData) {
-        return uri;
+    @Test
+    public void deleteFile_callsMarkOriginalFileOrDelete() {
+        FormEntryPrompt prompt = promptWithAnswer(new StringData(FILE_PATH));
+        when(prompt.getIndex()).thenReturn(formIndex);
+
+        AudioWidget widget = createWidget(prompt);
+        widget.deleteFile();
+
+        verify(mockedQuestionMediaManager).markOriginalFileOrDelete("questionIndex",
+                widget.getInstanceFolder() + File.separator + "blah.mp3");
     }
 
-    @Override
-    public void setUp() throws Exception {
-        super.setUp();
-        destinationName = RandomString.make();
+    @Test
+    public void clearAnswer_removesAnswerAndHidesPlayer() {
+        AudioWidget widget = createWidget(promptWithAnswer(new StringData(FILE_PATH)));
+        widget.clearAnswer();
+
+        assertThat(widget.getAnswer(), nullValue());
+        assertThat(audioHelper.isMediaStopped, equalTo(true));
+        verify(audioController).hidePlayer();
     }
 
-    @Override
-    protected void prepareForSetAnswer() {
-        when(formEntryPrompt.isReadOnly()).thenReturn(false);
+    @Test
+    public void clearAnswer_callsMarkOriginalFileOrDelete() {
+        FormEntryPrompt prompt = promptWithAnswer(new StringData(FILE_PATH));
+        when(prompt.getIndex()).thenReturn(formIndex);
 
-        when(mediaUtil.getPathFromUri(any(Context.class), any(Uri.class), any(String.class)))
-                .thenReturn(String.format("%s.mp3", RandomString.make()));
+        AudioWidget widget = createWidget(prompt);
+        widget.clearAnswer();
 
-        when(fileUtil.getRandomFilename()).thenReturn(destinationName);
+        verify(mockedQuestionMediaManager).markOriginalFileOrDelete("questionIndex",
+                widget.getInstanceFolder() + File.separator + "blah.mp3");
+    }
 
-        File firstFile = mock(File.class);
+    @Test
+    public void clearAnswer_callsValueChangeListeners() {
+        AudioWidget widget = createWidget(promptWithAnswer(new StringData(FILE_PATH)));
+        WidgetValueChangedListener valueChangedListener = mockValueChangedListener(widget);
+        widget.clearAnswer();
 
-        when(fileUtil.getFileAtPath(File.separator + destinationName + ".mp3"))
-                .thenReturn(firstFile);
+        verify(valueChangedListener).widgetValueChanged(widget);
+    }
+
+    @Test
+    public void setData_updatesWidgetAnswer() {
+        AudioWidget widget = createWidget(promptWithAnswer(new StringData(FILE_PATH)));
+        widget.setBinaryData(mockedFile);
 
-        when(firstFile.exists()).thenReturn(true);
-        when(firstFile.getName()).thenReturn(destinationName);
+        assertThat(widget.getAnswer().getDisplayText(), equalTo(FileWidgetUtils.updateWidgetAnswer(widgetActivity, mockedFile, "questionIndex",
```

Comment:
Also, I wanted to ask, is it fine if just check whether the answer updated here is the same as returned by `FileWidgetUtils.updateWidgetAnswer()`? I'll have to write separate tests for `FileWidgetUtils` class

Reply:
I think it's a sign that you maybe want a different interface/object that's responsible for computing the answer? That or you just want to have a hardcoded expectation. I'd try and avoid writing tests that assert something is the same as the result of something else.